app/models/collection_number.rb

```diff
@@ -102,25 +107,68 @@ def log_destroy
   end
 
   def destroy_without_callbacks
-    self.class.connection.execute(%(
-      DELETE FROM collection_numbers WHERE id = #{id}
-    ))
+    delete_manager = arel_delete_collection_numbers
+    self.class.connection.execute(delete_manager.to_sql)
   end
 
+  private
+
+  # DELETE FROM collection_numbers WHERE id = #{id}
+  def arel_delete_collection_numbers
+    cn = Arel::Table.new(:collection_numbers)
+    Arel::DeleteManager.new.from(cn).where(cn[:id].eq(id))
+  end
+
+  public
+
   # Mirror changes to collection number in herbarium records.  Do this
   # low-level to avoid redundant rss logs and other callbacks.
   def change_corresponding_herbarium_records(old_format_name)
-    new_format_name = Observation.connection.quote(format_name)
-    old_format_name = Observation.connection.quote(old_format_name)
-    Observation.connection.execute(%(
-      UPDATE collection_numbers_observations cno,
-             herbarium_records_observations hro,
-             herbarium_records hr
-      SET hr.accession_number = #{new_format_name}
-      WHERE cno.collection_number_id = #{id}
-        AND cno.observation_id = hro.observation_id
-        AND hro.herbarium_record_id = hr.id
-        AND hr.accession_number = #{old_format_name}
-    ))
+    update_manager = arel_update_corresponding_herbarium_records(
+      old_format_name
+    )
+    # puts(update_manager.to_sql)
+    Observation.connection.execute(update_manager.to_sql)
+  end
+
+  private
+
+  # UPDATE collection_numbers_observations cno,
+  #   herbarium_records_observations hro,
+  #   herbarium_records hr
+  # SET hr.accession_number = #{new_format_name}
+  #   WHERE cno.collection_number_id = #{id}
+  #   AND cno.observation_id = hro.observation_id
+  #   AND hro.herbarium_record_id = hr.id
+  #   AND hr.accession_number = #{old_format_name}
+  def arel_update_corresponding_herbarium_records(old_format_name)
```

Comment:
Yikes!  This method cannot be called very often.  It was kinda ugly to start with, but now it's insane and impossible to fix the ABC complexity.  What would you think about rewriting this entirely in ActiveRecord?

  observaton.herbarium_records.where(accession_number: old_format_name).update_all(accession_number: format_name)

(I'm trying to look up the API dock for update_all to make sure I'm remember the syntax right, but it won't load.  Sorry.  But update_all definitely doesn't fire callbacks, so it should be the right one to use, at least.)

Reply:
Was just remembering this!
I don't believe what i have here can even work, since it is also an UpdateManager on a join.
`update_all` it is!

Reply:
Here's what API dock has for examples:
```
# Update all customers with the given attributes
[Customer](https://apidock.com/rails/Customer).[update_all](https://apidock.com/rails/ActiveRecord/Relation/update_all) wants_email: true

# Update all books with 'Rails' in their title
Book.where('title LIKE ?', '%Rails%').[update_all](https://apidock.com/rails/ActiveRecord/Relation/update_all)(author: 'David')

# Update all books that match conditions, but limit it to 5 ordered by date
Book.where('title LIKE ?', '%Rails%').order(:created_at).limit(5).[update_all](https://apidock.com/rails/ActiveRecord/Relation/update_all)(author: 'David')

# Update all invoices and set the number column to its id value.
Invoice.[update_all](https://apidock.com/rails/ActiveRecord/Relation/update_all)('number = id')

Model.where(:foo => 'bar').where(:attr => 1).update_all("author = 'David'")
```

Reply:
Completely rewrote it.